[ZShunRen] iP

ZShunRen opened the pull request:
# Ned
> "When the snows fall and the white winds blow, the lone wolf dies, but the pack survives" - Ned Stark  ([source](https://awoiaf.westeros.org/index.php/Eddard_Stark#cite_note-Ragot22.7B.7B.7B3.7D.7D.7D.7B.7B.7B4.7D.7D.7D-54))
Ned may be a lord, but he's here to serve you! He will help you remember things to do, for the North remembers. He's
- text-based
- simple and blunt
- ~~honourable~~ fast at his duties
All you need to summon Ned and utilise his services is:
1. Download him from [here](https://github.com/ZShunRen/ip/releases).
2.  Run the following command in the folder where you downloaded him:
```ps
java -jar Ned.java
```
3. Add tasks.
4. Leave the rest to Ned. Ned is now **forever** sworn to aid you for the rest of his days! :crossed_swords:

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon) 

RIght now the `main` method looks like this:
```java
    public static void main(String[] args) {
        new Ned(Ned.cachedTasksPath).run();
    }
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

ZShunRen committed:
Add new name print

ZShunRen committed:
Add level 0

ZShunRen committed:
Add level 1

ZShunRen committed:
Add level 2

ZShunRen committed:
Add level 3

ZShunRen committed:
Make Task class abstract, created 3 subtypes of task

ZShunRen committed:
Add level 4

ZShunRen committed:
Add level 5

ZShunRen committed:
Add level 6

ZShunRen committed:
Remove horizontal lines

ZShunRen committed:
Add more tests

ZShunRen committed:
Update Readme.md

ZShunRen committed:
Refactor Ned class, by moving command logic into separate classes

ZShunRen committed:
Refactor CommandManager

ZShunRen committed:
Add enums in CommandManager + refactored code to meet standard

ZShunRen committed:
remove unnecessary new lines + fix expected test output

ZShunRen committed:
Refactor execution process and add caching using local data file

ZShunRen committed:
refactor command manager class, splitting responsibility to different classes, added abstract command class + added constant to cache file path

ZShunRen committed:
Merge branch 'branch-Level-7'

ZShunRen committed:
Add minimal version of Level 8

ZShunRen committed:
Refactor Ned and added Storage class to deal with loading and saving of results to cachefile

ZShunRen committed:
Refactor code into packages + add Ui class

ZShunRen committed:
Add tasklist class, will refactor rest later

ZShunRen committed:
Update runtest.bat logic and finished refactoring

ZShunRen committed:
Finish refactor as according to moreoop specs

ZShunRen committed:
Merge branch 'MoreOOP'

ZShunRen committed:
Renamed packages to fit spec + add data folder to prevent errors with caching

ZShunRen committed:
Add data folder + empty cache file to prevent errors from not being able to write to a non-existent directory

ZShunRen committed:
Merge branch 'add-gradle-support' of https://github.com/ZShunRen/ip into add-gradle-support

ZShunRen committed:
Add gradle support

ZShunRen committed:
Add junit test for ToDo class

ZShunRen committed:
Add Junit tests for two methods in ToDo and Deadline classes

ZShunRen committed:
Add javadocs for public methods in AddDeadlineCommand class

ZShunRen committed:
Refactor CommandManager to Parser to fit spec

ZShunRen committed:
Refactor CommandManager to Parser to fit spec

ZShunRen committed:
Merge branch 'master' of https://github.com/ZShunRen/ip into branch-Level-8

ZShunRen committed:
Refactor of CommandManager to Parser complete and done in other files

ZShunRen committed:
Refactor of CommandManager to Parser complete and done in other files

ZShunRen committed:
Merge branch 'master' of https://github.com/ZShunRen/ip into A-JavaDoc

ZShunRen committed:
Remove FlagWrapper class and add javadoc comments for all public methods

ZShunRen committed:
Remove FlagWrapper class and added javadoc comments for all public methods

ZShunRen committed:
Refactor code to be inline with standards

ZShunRen committed:
Refactor code to be inline with standards

ZShunRen committed:
Add level 9 and removed ned.task mention in comments

ZShunRen committed:
Add level 9 and remove ned.task mention in comments

ZShunRen committed:
Merge branch 'master' of https://github.com/ZShunRen/ip into A-Jar

aldentantan commented on src/main/java/ned/Storage.java:
```diff
@@ -0,0 +1,80 @@
+
+public class Storage {
+
+    private File cacheFile;
+
+    private String cacheFilePath;
```
Inconsequential but the blank line between attributes really stood out to me

ZShunRen replied:
Thank you for the pointer, I have since removed these blank lines as I realised that they don't subscribe to the coding standard

aldentantan commented on src/main/java/ned/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+
+    public void addTask(Task newTask, Ui uiInstance) {
+        this.listOfTasks.add(newTask);
+        uiInstance.print("Aye, I've added this task m'lord:");
+        uiInstance.print(Ui.INDENTATIONS + newTask);
+        uiInstance.print("Now you've " + this.listOfTasks.size() + " tasks left. Get to it then!");
```
Is there a way to get Ui to handle all the printing? Maybe a `uiInstance.printAddTask()` function

ZShunRen replied:
Yes I have added a method that will handle this more succinctly, thank you

aldentantan commented on src/main/java/ned/commands/AddEventCommand.java:
```diff
@@ -0,0 +1,70 @@
+package ned.commands;
+
+import ned.Storage;
+import ned.TaskList;
+import ned.exceptions.NedException;
+
+import ned.tasks.Task;
+import ned.tasks.Event;
+
+import ned.Ui;
+
+import java.util.regex.Pattern;
```
Can standardise the format of import statements, whether to have blank lines in between or not

ZShunRen replied:
Thanks for the pointer, I also didn't realise that I had this inconsistent formatting

aldentantan reviewed (commented):
Looks good to me coding standard wise :)

ZShunRen committed:
Add Level 10

ZShunRen committed:
Fix code style issues with spacing

ZShunRen committed:
Reformat code + add method to address feedback

ZShunRen committed:
Address first round of checkstyle issues

ZShunRen committed:
Add spacing between import statements

shadhankkk commented on src/main/java/ned/tasks/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+package ned.tasks;
```
I like that your packages are in all lower case

ZShunRen replied:
thanks!

shadhankkk commented on src/main/java/ned/tasks/Deadline.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    @Override
+    public boolean equals(Object obj) {
```
Maybe this method name can be renamed to something that follows the is, has, was, etc. format (e.g isEquals)

ZShunRen replied:
fair point, but issue here is that I am overriding the 'equals' method in the object class of java

shadhankkk reviewed (commented):
Great code overall!

ZShunRen committed:
Add new exception subclasses

ZShunRen committed:
Add new exception subclasses

There exists only the NedException exception class.

There are now several subclasses of NedException, with self-descriptive
names to describe the error that occured. They wrap around
some run-time exceptions and provide more information in the
context of the chatbot.

Adding these subclasses makes the code easier to understand.

ZShunRen committed:
Assign error messages to constants where possible

Most of the error messages are passed to the exception constructor
directly.

Extracting these message into constants makes the code more readable as
error messages that pertain to that specific exception.

This removes the magic literal strings, and makes it easier to
understand at a glance.

However, not all error messages could be extracted into constants so
easily, especially those that use method parameters in their lines.

This applies to Parser.java at this point of time.

ZShunRen committed:
Add assert statements in key parts of code

Some key assumptions, like that the list of tasks exists, and is not
null, are not tested.

By testing these key assumptions, although they may hold true for now,
when doing future refactoring or feature-adding, these assumptions may
break.

These new assert statements help to ensure that this is caught.

ZShunRen committed:
Add detection of duplicate task check

Tasks are added to the tasklist, without checking if they already exist.

It is important to ensure tasks, which are identical in their nature,
are not added, as this can confuse the user.

The chatbot will inform the user if the task they are trying to
add, already exists in the list of tasks.

ZShunRen committed:
Merge branch 'master' into A-CodeQuality

ZShunRen committed:
Resolve merge conflict from merging A-CodeQuality

ZShunRen committed:
Resolve merge conflict from pr to branch

ZShunRen committed:
Merge pull request #2 from ZShunRen/A-CodeQuality

 Assign error messages to constants where possible

ZShunRen committed:
Resolve merge of A-Assertions with master

ZShunRen committed:
Merge pull request #1 from ZShunRen/A-Assertions

Add assert statements in key parts of code

ZShunRen committed:
Add gradle yml file for CI

ZShunRen committed:
Fix styling issues with checkstyle

ZShunRen committed:
Fix all checkstyle issues

ZShunRen committed:
Fix checkstyle error in test files

ZShunRen committed:
Improve UI and add new help command

Add the following changes to UI:
- new House Stark background
- colourings for dialog boxes depending if it's user or ned
- red colouring for dialog box if it's an error message
- dynamic resizing of window
- prompt text to user input
- change send button message to 'decree'
Add new help command, which is activated by user inputting 'help'.

The UI was generic and lacking of a theme suited to the personality of
its chatbot personality.
A help command to guide users on which commands and how to use them is
not present.

ZShunRen committed:
Add creditation in README.md

ZShunRen committed:
Add logo and replace copyrighted display picture

ZShunRen committed:
Merge tag 'A-BetterGUI'

ZShunRen committed:
Merge branch 'refs/heads/A-BetterGUI'

ZShunRen committed:
Amend remove task message

ZShunRen committed:
Update command message

ZShunRen committed:
Fix missing mark command in help message

ZShunRen committed:
Move cache file to data subdirectory of root

ZShunRen committed:
Amended data subfolder path

ZShunRen committed:
Merge detect duplicates with master

ZShunRen committed:
Complete merge with detect duplicates

ZShunRen committed:
Merge with A-CheckStyle

ZShunRen committed:
Add Ai.md file to project root

ZShunRen committed:
Merge tag 'A-AiAssisted'

ZShunRen committed:
Update readme.md in docs subfolder

ZShunRen committed:
Merge tag 'A-UserGuide'

ZShunRen committed:
Add Ui picture back

ZShunRen committed:
Merge branch 'refs/heads/A-UserGuide'

ZShunRen committed:
Add message for empty list and resolve code style

Ned shows an empty message body if the list of tasks is empty.
In addition there are several code style issues like a lack of
standardisation of where the exit message is derived from and formatting
issues in the docs/README.md file.

- When `list` command is run and if the size of the listOfTasks == 0,
then a special message is shown to the user to indicate the list is
empty.
- The exit message has been refactored to a static constant belonging to Ned class,
with a public getter method for future access to it.
- Broken links in the README.md file have been removed, and formatting
  issues resolved.
